router/quotation.gno

```diff
@@ -0,0 +1,184 @@
+		}
+	}
+
+	// DrySwap to calculate
+	for i, quoterTarget := range quoterTargets {
+		numPools := strings.Count(quoterTarget.targetPath, ",") / 2
+		require(numPools >= 1 && numPools <= 3, ufmt.Sprintf("[ROUTER] quotation.gno__quoteForAllPath() || numPools should 1 ~ 3, but found %d", numPools))
+
+		if numPools == 1 {
+			input, output, fee := getSwapData(quoterTarget.targetPath, 0)
+
+			singleParams := SingleSwapParams{
+				tokenIn:           input,
+				tokenOut:          output,
+				fee:               fee,
+				amountSpecified:   quoterTarget.pctAmount,
+				sqrtPriceLimitX96: sqrtPriceLimitX96,
+			}
+
+			estimatedResult := singleSwapDry(singleParams)
+			if estimatedResult > 0 {
+				if amountSpecified > 0 {
+					resultRatioX96 := estimatedResult * Q96 / quoterTarget.pctAmount * Q96 / Q96
+					quoterTarget.resultRatioX96 = resultRatioX96
+					quoterTargets[i] = quoterTarget
+				} else {
+					resultRatioX96 := -quoterTarget.pctAmount * Q96 / estimatedResult * Q96 / Q96
+					quoterTarget.resultRatioX96 = resultRatioX96
+					quoterTargets[i] = quoterTarget
+				}
+			} else {
+				panic("[ROUTER] quotation.gno__quoteForAllPath() || SINGLE__estimateResult < 0")
+			}
+		}
+
+		if numPools > 1 && numPools <= 3 {
+			input, output, fee := getSwapData(quoterTarget.targetPath, 0)
+
+			swapParams := SwapParams{
+				tokenIn:  input,
+				tokenOut: output,
+				fee:      fee,
+
+				recipient:       std.GetOrigCaller(),
+				amountSpecified: quoterTarget.pctAmount,
+				minAmountOut:    1, // r3v4_xx: sqrtPriceLimitX96
+			}
+
+			// if amountSpecified > 0, proceed forward
+			// if not, proceed with backward
+			if amountSpecified > 0 {
+				estimatedResult := multiSwapDry(swapParams, 0, numPools, quoterTarget.targetPath) // will iterate here to cover multi pools
+				if estimatedResult > 0 {
+					resultRatioX96 := estimatedResult * Q96 / quoterTarget.pctAmount * Q96 / Q96
+					quoterTarget.resultRatioX96 = resultRatioX96
+					quoterTargets[i] = quoterTarget
+				} else {
+					panic("[ROUTER] quotation.gno__quoteForAllPath() || MULTI__PositiveAmountSpecified__estimateResult < 0")
+				}
+			} else {
+				estimatedResult := multiSwapNegativeDry(swapParams, numPools-1, quoterTarget.targetPath) // will iterate here to cover multi pools
+				if estimatedResult > 0 {
+					resultRatioX96 := -quoterTarget.pctAmount * Q96 / estimatedResult * Q96 / Q96
+
+					quoterTarget.resultRatioX96 = resultRatioX96
+					quoterTargets[i] = quoterTarget
+				} else {
+					panic("[ROUTER] quotation.gno__quoteForAllPath() || MULTI__NegativeAmountSpecified__estimateResult < 0")
+				}
+			}
+
+		}
+	}
+
+	// it contains outputRatio with 0 which is impossible path
+	// > remove path with 0 ratio
+	var finalTargets []QuoterTarget
+	for _, quote := range quoterTargets {
+		if quote.resultRatioX96 != 0 {
+			finalTargets = append(finalTargets, quote)
+		}
+	}
+
+	sort.Sort(ByOutputRatioDesc(finalTargets))
```

Comment:
It's not critical, but it would be nice to only sort if the slice is non-empty.

Reply:
lets sort when len finalTagets > 0